src/lib.rs

```diff
@@ -0,0 +1,253 @@
+    unused_results,
+    missing_copy_implementations
+)]
+// This one is hard to avoid.
+#![allow(clippy::multiple_crate_versions)]
+
+pub mod auth;
+mod core;
+
+use crate::core::OperationHandler;
+use auth::AuthToken;
+use parsec_interface::operations::list_opcodes::Operation as ListOpcodes;
+use parsec_interface::operations::list_providers::{Operation as ListProviders, ProviderInfo};
+use parsec_interface::operations::ping::Operation as Ping;
+use parsec_interface::operations::psa_algorithm::AsymmetricSignature;
+use parsec_interface::operations::psa_destroy_key::Operation as PsaDestroyKey;
+use parsec_interface::operations::psa_export_public_key::Operation as PsaExportPublicKey;
+use parsec_interface::operations::psa_generate_key::Operation as PsaGenerateKey;
+use parsec_interface::operations::psa_import_key::Operation as PsaImportKey;
+use parsec_interface::operations::psa_key_attributes::KeyAttributes;
+use parsec_interface::operations::psa_sign_hash::Operation as PsaSignHash;
+use parsec_interface::operations::psa_verify_hash::Operation as PsaVerifyHash;
+use parsec_interface::operations::{NativeOperation, NativeResult};
+use parsec_interface::requests::Result;
+use parsec_interface::requests::{Opcode, ResponseStatus};
+use std::collections::HashSet;
+
+pub use parsec_interface::requests::ProviderID as Provider;
+
+/// Core client for Parsec service
+///
+/// The client exposes low-level functionality for using the Parsec service
+#[derive(Debug)]
+pub struct CoreClient {
+    op_handler: OperationHandler,
+    auth_token: AuthToken,
+}
+
+impl CoreClient {
+    /// Create a new client
+    pub fn new(auth_token: AuthToken) -> Self {
+        CoreClient {
+            op_handler: Default::default(),
+            auth_token,
+        }
+    }
+
+    /// list opcodes
+    pub fn list_provider_operations(&self, provider: Provider) -> Result<HashSet<Opcode>> {
+        let res = self.op_handler.send_operation(
+            NativeOperation::ListOpcodes(ListOpcodes {}),
+            provider,
+            &self.auth_token,
+        )?;
+
+        if let NativeResult::ListOpcodes(res) = res {
+            Ok(res.opcodes)
+        } else {
+            Err(ResponseStatus::InvalidHeader)
+        }
+    }
+
+    /// list providers
+    pub fn list_providers(&self) -> Result<Vec<ProviderInfo>> {
+        let res = self.op_handler.send_operation(
+            NativeOperation::ListProviders(ListProviders {}),
+            Provider::Core,
+            &self.auth_token,
+        )?;
+
+        if let NativeResult::ListProviders(res) = res {
+            Ok(res.providers)
+        } else {
+            Err(ResponseStatus::InvalidHeader)
+        }
+    }
+
+    /// ping
+    pub fn ping(&self) -> Result<()> {
+        let _ = self.op_handler.send_operation(
+            NativeOperation::Ping(Ping {}),
+            Provider::Core,
+            &AuthToken::Null,
+        )?;
+
+        Ok(())
+    }
+
+    /// generate key
+    pub fn generate_key(
```

Comment:
Should we prefix them with `psa_` as well?

Reply:
Don't know what to say. Maybe at this lowest layer it could make sense, but I don't think clients would be particularly interested in the fact that this operation comes from the PSA spec. Might be wrong... I think my point is that this could(/should?) be an abstraction over the things offered by the service. Not sure how it would be settled for a client that supports some provider-specific version of, say, `generate_key`

Reply:
> Maybe at this lowest layer it could make sense
That is what I was thinking as well: that this low layer (the "core") could be the dumb translation, in client side, of all our operation with all parameters. With that in mind, it would make sense to me to have the `psa_` prefix, as those are `PsaXXX` operations.

I agree that once on higher abstraction layers, we should probably drop the prefix as it would make less sense.